lib/Sema/CSSimplify.cpp

```diff
@@ -1417,13 +1392,53 @@ assessRequirementFailureImpact(ConstraintSystem &cs, Type requirementType,
   // fix for same-type requirement higher impact vs. requirement associated
   // with method itself e.g. `func foo<U>() -> U where U : P {}` because
   // `foo` is accessible from any `S` regardless of what `T` is.
+  //
+  // Don't add this impact with the others, as we want to keep it consistent
+  // across requirement failures to present the user with a choice.
   if (isa<UnresolvedDotExpr>(anchor) || isa<UnresolvedMemberExpr>(anchor)) {
     auto *calleeLoc = cs.getCalleeLocator(cs.getConstraintLocator(locator));
     if (!cs.findSelectedOverloadFor(calleeLoc))
       return 10;
   }
 
-  return 1;
+  // Increase the impact of a conformance fix for a standard library type,
+  // as it's unlikely to be a good suggestion. Also do the same for the builtin
+  // compiler types Any and AnyObject, which cannot conform to protocols.
+  // FIXME: We ought not to have the is<TypeVariableType>() condition here, but
+  // removing it currently regresses the diagnostic for the test case for
+  // rdar://60727310. Once we better handle the separation of conformance fixes
+  // from argument mismatches in cases like SR-12438, we should be able to
+  // remove it from the condition.
+  auto resolvedTy = cs.simplifyType(requirementType);
+  if ((requirementType->is<TypeVariableType>() && resolvedTy->isStdlibType()) ||
+      resolvedTy->isAny() || resolvedTy->isAnyObject()) {
+    if (auto last = locator.last()) {
+      if (auto requirement = last->getAs<LocatorPathElt::AnyRequirement>()) {
+        auto kind = requirement->getRequirementKind();
+        if (kind == RequirementKind::Conformance)
+          impact += 2;
+      }
+    }
+  }
+
+  // If this requirement is associated with an overload choice let's
+  // tie impact to how many times this requirement type is mentioned.
+  if (auto *ODRE = dyn_cast<OverloadedDeclRefExpr>(anchor)) {
```

Comment:
What do you think about an idea of not considering requirement failures recoverable in reference to operators? 

I'm just trying to figure out how to move a needle in situations like:

```swift
func sr12438(_ e: Error) {
  func foo<T>(_ a: T, _ op: ((T, T) -> Bool)) {}
  foo(e, ==) // expected-error {{type of expression is ambiguous without more context}}
}
```

The problem here is that none of the operator declarations for `==` are going to match and they are all going to fail in different ways. I think we shouldn't try to rank such situations but instead be better at detecting completely hopeless situations...

Reply:
Hmm, would that be specifically for unapplied operator references, or for operator references in general? Presumably we'd then emit a generic "operator cannot be applied with X operands" diagnostic?

I guess we could just try to detect the case where we have different possible overloads for an operator in `diagnoseAmbiguityWithFixes`, each with some kind of fix, and call into `diagnoseOperatorAmbiguity` (though it may need to be tweaked slightly to handle unapplied operator refs).

Reply:
> Hmm, would that be specifically for unapplied operator references, or for operator references in general?

In general. I am hoping that improvement is going to be two-fold - fewer operators considered in diagnostics mode and more precise diagnostics when no operator overloads match.

That's what I was thinking as well, the problem is that we need to move `diagnoseOperatorAmbiguity` before `findBestSolution` and that could affect some of diagnostics.

Reply:
> In general. I am hoping that improvement is going to be two-fold - fewer operators considered in diagnostics mode and more precise diagnostics when no operator overloads match.

My only worry is for people defining their own operators that require conformances for arguments, it would be a shame to lose the requirement fixes there, perhaps we should limit it to known stdlib operators? Or maybe operators that have lots of overloads?

Reply:
Yeah, maybe that should be limited only to stdlib operator overloads only.